plugins/ecs-serviceconnect/plugin/commands.go

```diff
@@ -0,0 +1,244 @@
+	// Names of iptables chains created for ECS Service Connect rules.
+	ingressChain = "ECS_SERVICE_CONNECT_INGRESS"
+	egressChain  = "ECS_SERVICE_CONNECT_EGRESS"
+)
+
+// Add is the internal implementation of CNI ADD command.
+func (plugin *Plugin) Add(args *cniSkel.CmdArgs) error {
+	// Parse network configuration.
+	netConfig, err := config.New(args)
+	if err != nil {
+		log.Errorf("Failed to parse netconfig from args: %v.", err)
+		return err
+	}
+	log.Infof("Executing ADD with netconfig: %+v.", netConfig)
+
+	// Find the network namespace.
+	log.Debugf("Searching for netns %s.", args.Netns)
+	ns, err := netns.GetNetNS(args.Netns)
+	if err != nil {
+		log.Errorf("Failed to find netns %s: %v.", args.Netns, err)
+		return err
+	}
+	// Add IP rules in the target network namespace.
+	err = ns.Run(func() error {
+		for _, proto := range netConfig.IPProtocols {
+			if err = plugin.setupIptablesRules(proto, netConfig); err != nil {
+				log.Errorf("Failed to set up iptables rules: %v.", err)
+				return err
+			}
+		}
+		return nil
+	})
+	// TODO: Do we need to return CNI result?
+	return err
+}
+
+// setupIptablesRules sets iptables/ip6tables rules in container network namespace.
+func (plugin *Plugin) setupIptablesRules(proto iptables.Protocol, config *config.NetConfig) error {
+	// Create a new iptables object.
+	iptable, err := iptables.NewWithProtocol(proto)
+	if err != nil {
+		return err
+	}
+	// Setup Ingress rules
+	if err = plugin.setupIngressRules(iptable, config); err != nil {
+		return err
+	}
+	// Setup Egress rules
+	return plugin.setupEgressRules(iptable, proto, config)
+}
+
+// setupIngressRules sets Ingress port redirection rules in the network namespace
+func (plugin *Plugin) setupIngressRules(
+	iptable *iptables.IPTables,
+	config *config.NetConfig) error {
+	// Skip setting rules when there is no redirection
+	if len(config.IngressRedirectToListenerPortMap) == 0 {
+		return nil
+	}
+
+	// Redirect non-local traffic to the redirection port
+	return pluginutils.RedirectNonLocalTraffic(iptable, ingressChain, config.IngressRedirectToListenerPortMap)
+}
+
+// setupEgressRules sets Egress port redirection rules in the network namespace
+func (plugin *Plugin) setupEgressRules(
+	iptable *iptables.IPTables,
+	proto iptables.Protocol,
+	config *config.NetConfig) error {
+	// Skip setting rules when there is no egress listener port
+	if config.EgressPort == 0 {
+		return nil
+	}
+
+	// Redirect traffic in the CIDR block to the egress port
+	return plugin.redirectCIDRTraffic(iptable, config.Mode, plugin.getCidr(proto, config), strconv.Itoa(config.EgressPort))
+}
+
+// redirectCIDRTraffic sets rules to redirect traffic in the CIDR block to the egress port
+func (plugin *Plugin) redirectCIDRTraffic(
+	iptable *iptables.IPTables,
+	mode config.NetworkMode,
+	cidr string,
+	redirectPort string) error {
+	var err error
+	switch mode {
+	case config.AWSVPC, config.BRIDGE_DNAT:
+		err = iptable.Append("nat", "OUTPUT", "-p", "tcp", "-d", cidr,
```

Comment:
Thanks for pointing it out. I will pick this up in the next PR. Currently modifying it to only do redirection of CIDRs so that AwsVPC mode can be tested out.